src/librbd/mirror/snapshot/GroupPrepareImagesRequest.cc

```diff
@@ -0,0 +1,426 @@
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::handle_notify_quiesce(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0 &&
+      (m_internal_flags & SNAP_CREATE_FLAG_IGNORE_NOTIFY_QUIESCE_ERROR) == 0) {
+    m_ret_code = r;
+    notify_unquiesce();
+    return;
+  }
+
+  acquire_image_exclusive_locks();
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::acquire_image_exclusive_locks() {
+  ldout(m_cct, 10) << dendl;
+
+  auto ctx = librbd::util::create_context_callback<
+    GroupPrepareImagesRequest<I>,
+    &GroupPrepareImagesRequest<I>::handle_acquire_image_exclusive_locks>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (auto ictx: *m_image_ctxs) {
+    std::shared_lock owner_lock{ictx->owner_lock};
+    if (ictx->exclusive_lock != nullptr) {
+      ictx->exclusive_lock->block_requests(-EBUSY);
+      ictx->exclusive_lock->acquire_lock(gather_ctx->new_sub());
+    }
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::handle_acquire_image_exclusive_locks(
+    int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to acquire image exclusive locks: "
+                 << cpp_strerror(r) << dendl;
+    m_ret_code = r;
+    if (!m_quiesce_requests->empty()) {
+      notify_unquiesce();
+    } else {
+      remove_snap_metadata();
+    }
+    return;
+  }
+
+  finish(0);
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::notify_unquiesce() {
+  ldout(m_cct, 10) << dendl;
+
+  ceph_assert(m_quiesce_requests->size() == m_image_ctxs->size());
+
+  auto ctx = librbd::util::create_context_callback<
+    GroupPrepareImagesRequest<I>,
+    &GroupPrepareImagesRequest<I>::handle_notify_unquiesce>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  int image_count = m_image_ctxs->size();
+  for (int i = 0; i < image_count; ++i) {
+    auto ictx = (*m_image_ctxs)[i];
+    ictx->image_watcher->notify_unquiesce((*m_quiesce_requests)[i],
+                                          gather_ctx->new_sub());
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::handle_notify_unquiesce(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    // ignore error
+    lderr(m_cct) << "failed to notify the unquiesce requests: "
+                 << cpp_strerror(r) << dendl;
+  }
+
+  remove_snap_metadata();
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::remove_snap_metadata() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = cls_client::group_snap_remove(
+    &m_group_ioctx, librbd::util::group_header_name(m_group_id),
+    m_group_snap->id);
+  if (r < 0) {
+    // ignore error
+    lderr(m_cct) << "failed to remove group snapshot metadata: "
+                 << cpp_strerror(r) << dendl;
+  }
+
+  close_images();
+}
+
+template <typename I>
+void GroupPrepareImagesRequest<I>::close_images() {
+  ldout(m_cct, 10) << dendl;
+
```

Comment:
Check if the group is empty and finish here.

Reply:
Done. Added this
`  if (m_image_ctxs->empty()) {
    finish(m_ret_code);
    return;
  }
`